### PR TITLE
Support for parsing gregorian dates

### PR DESCRIPTION
**Pull request by jystic**
@charleso I benchmarked this against the `parseDay` I sent you and it's 50% faster again (40ns instead of 60ns per call).

! @charleso @amosr @tranma

### CODE REVIEW

**Comment by jystic** [comment]
Spent some time optimising and adding validation, it's now down to about 30ns per call and validates whether the parsed date is in fact in the gregorian calendar.

I tested this on an AWS box and the results aren't quite as good as locally. Aside from the hardware differences, GCC isn't able to get rid of the division instructions in the final assembly. LLVM does a great job with this on my mac, not a single `idiv` instruction, every division is turned in to something else.

Speed wise on AWS it takes about 80ns per `parseDay` call, where as `parseYearMonthDay` only takes about 40ns per call. I put this down to the gregian -> julian day conversion which requires quite a few divisions.

When the month is February, the `parseDay` call jumps to 100ns, and the `parseYearMonthDay` jumps to 60ns, as we need to do a leap year check and that involves further divisions which GCC fails to optimise.

*My local machine / laptop:*
Local

*AWS box:*
AWS

**Comment by amosr** [comment]
bit confused by the benchmark because they're returning different things. if you were to use `parseYearMonthDay` in a parser would you end up converting it to a `Day` in Haskell?

File: test/Test/Anemone/Foreign/Time.hs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        fmap (, Char8.pack xs) $ mkDay ymd
+
+      str =
+        Char8.pack $ printf "%04d-%02d-%02d%s" y m d xs
+    in
+      parseDay str === eday
+
+return []
+tests =
+  $forAllProperties $ quickCheckWithResult (stdArgs {maxSuccess = 10000}) -- 10k certified
```
Comment (amosr) [inline test/Test/Anemone/Foreign/Time.hs]: `$disorderCheckEnvAll TestRunMore`

https://github.com/ambiata/anemone/blob/master/test/Test/Anemone/Foreign/Memcmp.hs#L63

**Comment by amosr** [comment]
real groovy 👍

**Comment by jystic** [comment]
> bit confused by the benchmark because they're returning different things. if you were to use `parseYearMonthDay` in a parser would you end up converting it to a `Day` in Haskell?

Talked IRL, you might want to use `parseYearMonthDay` if you just want to validate that something is a date, and you don't care about the result. With clang/llvm there is no difference, but on gcc1970 `parseYearMonthDay` is twice as fast.

**Pull request merged**